MPI related packages: use multiple outputs

markuskowa opened the pull request:
## Description of changes
Separate  runtime file from documentation/man pages, and build time files (where possible).
Note that for the MPI packages the `dev` output cannot easily be separated as this leads to cyclic dependencies between `out` and `dev`.


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

markuskowa committed:
libfabric: split outputs -> out, dev, man

markuskowa committed:
mpich: split outputs -> out, doc, man

markuskowa committed:
mvapich: split outputs -> out, doc, man

markuskowa committed:
ucc: split outputs -> out, dev

markuskowa committed:
ucx: split outputs -> out, doc, dev

markuskowa commented:
@GrahamcOfBorg test slurm

bjornfor commented on pkgs/development/libraries/openmpi/default.nix:
```diff
@@ -66,13 +68,13 @@ in stdenv.mkDerivation rec {
 
   postInstall = ''
-    rm -f $out/lib/*.la
+    find $out/lib/ -name "*.la" -exec rm -f \{} \;
```
Unrelated change. Please put it in a separate commit and explain the what/why of that change.

(I checked and found that there is only one *.la file in the output: `result/lib/openmpi/libompi_dbg_msgq.la`. So that means the current code is a NOP. I'm not sure why we're removing *.la files though.)

markuskowa replied:
It actually has la files in `lib/` and the one in `lib/openmpi` (the old version just removed them):
```
result/lib/libmpi.la
result/lib/libmpi_usempif08.la
result/lib/libompitrace.la
result/lib/openmpi/libompi_dbg_msgq.la
result/lib/libmpi_usempi_ignore_tkr.la
result/lib/libmpi_mpifh.la
result/lib/liboshmem.la
result/lib/libopen-pal.la
result/lib/libmpi_cxx.la
result/lib/libopen-rte.la
```

A far as I am aware, these libtool files are not needed for anything on Linux (or on Darwin). The idea here is to simply to reduce the output size/number of files in the output.

bjornfor replied:
Hehe, silly me, the rm code worked :-)

bjornfor commented on pkgs/development/libraries/pmix/default.nix:
```diff
@@ -23,15 +23,19 @@ stdenv.mkDerivation rec {
   '';
 
+  postInstall = ''
+    find $out/lib/ -name "*.la" -exec rm -f \{} \;
```
Can you say why removing .la files is a good thing in the commit message?

markuskowa committed:
openmpi: split outputs -> out, man

markuskowa committed:
openmpi: use getDev to get includes

markuskowa committed:
openmpi: delete all libtool .la files

Clear all *.la from lib to reduce the number file in output.
These libtool files are not needed.

markuskowa committed:
pmix: use getDev to get includes

markuskowa committed:
pmix: remove libtool .la files

The libtool files are not needed. Clear the .la files from
the output to reduce the number files.

bzizou reviewed (approved):
LGTM
